### ISSUE TITLE
Bug: [no-empty-object-type] false positive/incorrect fixes for default type parameter

### ISSUE THREAD

**Issue by printfn** [typescript-eslint/typescript-eslint#9761]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.2&fileType=.tsx&code=KYDwDg9gTgLgBASwHY2FAZgQwMbDgSQB4AFOUVJAEwGc4AlYbaSw6mKZAcwBo42OknAHxwAvHADeAXxESAsACg4cMJiiYAttQBccYgG5FUoA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge3oFsbDaOAIwBWiMnSKl0URNGgdokcGAC%2BIFUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false

### Repro Code

```TypeScript
export interface I<P extends Record<string, string> = {}> {
  params: P;
}
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/no-empty-object-type": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

I'm not sure this should even produce an error. Neither of the suggested fixes works.

### Actual Result

The `{}` ("empty object") type allows any non-nullish value, including literals like `0` and `""`.
- If that's what you want, disable this lint rule with an inline comment or configure the 'allowObjectTypes' rule option.
- If you want a type meaning "any object", you probably want `object` instead.
- If you want a type meaning "any value", you probably want `unknown` instead. 5:55 - 5:57

> Replace `{}` with `object`.
> Replace `{}` with `unknown`.

**Comment by Josh-Cena** [comment]
The suggestions are there for common cases; neither of them are equivalent to `{}` and we don't guarantee that your code continues to work. In this case, the report is expected because it may result in unexpected behaviors:

```js
export interface I<P extends Record<string, string> = {}> {
  params: P;
}

const x: I = { params: 1 }; // Allowed
```

You can use `P extends Record<string, string> = Record<string, string>` as the default, or `P extends Record<string, string> = Record<string, never>` if you want the default to not contain any properties, or even just disable the rule, if you know what you are writing is intended.

**Comment by bradzacher** [comment]
Yeah this reporting is entirely correct and using `{}` here is unsafe as Josh mentions.
The suggestions are exactly that - suggestions. They aren't automatically applied and must be manually evaluated and manually applied.

In this case Josh's suggestion is correct - you have a generic constraint, so use the constraint as the default. Though you probably want something like `Record<string, never>` which would force it to be an empty object and accept `{ params: {} }` and reject `{ params: 1 }` and `{ params: { prop: 1 } }`.

---

@typescript-eslint/triage-team -- what do we think about adding a suggestion here:
If we detect that the `{}` is used as a generic default AND the generic has a constraint, then we additionally suggest the constraint as a special case suggestion

**Comment by Josh-Cena** [comment]
Hmmm, negative—I don't think that is going to help in most cases and really complicates the code.

**Comment by kirkwaiblinger** [comment]
-1 from me, too, I'd rather avoid blessing `Record<string, never>` with a typescript-eslint suggestion (see also https://github.com/typescript-eslint/typescript-eslint/pull/8977#discussion_r1578836006 and https://github.com/microsoft/TypeScript/issues/57735#issuecomment-1995395617)

**Comment by bradzacher** [comment]
I meant specifically if the code is of the form `<T extends U = {}>`
Then we would additionally suggest replacing `{}` with `U` (i.e. `<T extends U = U>`)

**Comment by Josh-Cena** [comment]
Yeah, I meant—this is not always, if ever, going to be what the user intends, and only solves a handful of cases out-of-the-box.

**Comment by kirkwaiblinger** [comment]
Oh, I misunderstood, thanks for clarifying.  +/- 0 from me on this 🤷‍♂️

**Comment by bradzacher** [comment]
Okay then it seems like we're in agreement - this is working as intended and no changes shall be made.
